s3_test.go

```diff
@@ -0,0 +1,272 @@
+		ExpectedKey:    "valid-key",
+		ExpectedBucket: "valid-bucket",
+	}
+
+	testCases := []struct {
+		Name           string
+		Key            string
+		Bucket         string
+		ExpectedResult bool
+		ExpectedErr    string
+	}{
+		{
+			Name:           "Valid inputs",
+			Key:            "valid-key",
+			Bucket:         "valid-bucket",
+			ExpectedResult: true,
+			ExpectedErr:    "<nil>",
+		},
+		{
+			Name:           "Invalid key",
+			Key:            "invalid",
+			Bucket:         "valid-bucket",
+			ExpectedResult: false,
+			ExpectedErr:    "<nil>",
+		},
+		{
+			Name:           "Invalid bucket",
+			Key:            "valid-key",
+			Bucket:         "invalid",
+			ExpectedResult: false,
+			ExpectedErr:    "<nil>",
+		},
+	}
+	for _, testCase := range testCases {
+		t.Run(testCase.Name, func(t *testing.T) {
+			result, err := ObjectExists(testCase.Key, testCase.Bucket)
+			if fmt.Sprintf("%+v", err) != testCase.ExpectedErr {
+				t.Fatalf("expected %+v but received %+v", testCase.ExpectedErr, err)
+			}
+			if result != testCase.ExpectedResult {
+				t.Fatalf("expected %t but received %t", testCase.ExpectedResult, result)
+			}
+		})
+	}
+}
+
+type mockedS3Session struct {
+	ExpectedKey    string
+	ExpectedBucket string
+}
+
+func (m *mockedS3Session) PutObject(ctx context.Context, params *s3.PutObjectInput, optFns ...func(*s3.Options)) (*s3.PutObjectOutput, error) {
+	if ctx == nil {
+		return nil, errors.New("context is nil")
+	}
+	if *params.Bucket != m.ExpectedBucket {
+		return nil, fmt.Errorf("unexpected bucket: %s", *params.Bucket)
+	}
+	if *params.Key != m.ExpectedKey {
+		return nil, fmt.Errorf("unexpected key: %s", *params.Key)
+	}
+	return &s3.PutObjectOutput{}, nil
+}
+
+func (m *mockedS3Session) GetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error) {
+	if ctx == nil {
+		return nil, errors.New("context is nil")
+	}
+	if *params.Bucket != m.ExpectedBucket {
+		return nil, fmt.Errorf("unexpected bucket: %s", *params.Bucket)
+	}
+	if *params.Key != m.ExpectedKey {
+		return nil, fmt.Errorf("unexpected key: %s", *params.Key)
+	}
+	return &s3.GetObjectOutput{}, nil
+}
+
+func (m *mockedS3Session) DeleteObject(ctx context.Context, params *s3.DeleteObjectInput, optFns ...func(*s3.Options)) (*s3.DeleteObjectOutput, error) {
+	if ctx == nil {
+		return nil, errors.New("context is nil")
+	}
+	if *params.Bucket != m.ExpectedBucket {
+		return nil, fmt.Errorf("unexpected bucket: %s", *params.Bucket)
+	}
+	if *params.Key != m.ExpectedKey {
+		return nil, fmt.Errorf("unexpected key: %s", *params.Key)
+	}
+	return &s3.DeleteObjectOutput{}, nil
+}
+
+func (m *mockedS3Session) HeadObject(ctx context.Context, params *s3.HeadObjectInput, optFns ...func(*s3.Options)) (*s3.HeadObjectOutput, error) {
+	if ctx == nil {
+		return nil, errors.New("context is nil")
+	}
+	if *params.Bucket != m.ExpectedBucket {
+		// TODO is this correct? There is also a types.NoSuchBucket in the AWS SDK.
```

Comment:
It might be a bug that we're not also looking for a `NoSuchBucket` error in the `ObjectExists` method, but since I'm trying to do the absolute minimum code change, I'm leaving out testing for it and potentially fixing it, because that would be in a separate PR.

Reply:
Follow-up, I did find that this isn't what AWS returns when a bucket or file is missing. After this PR is accepted and merged, I have an immediate follow-on PR I can post that adds _even more_ tests and fixes this! 😄